zookeeper-server/src/main/java/org/apache/zookeeper/server/NIOServerCnxnFactory.java

```diff
@@ -820,7 +820,12 @@ private void addCnxn(NIOServerCnxn cnxn) throws IOException {
     }
 
     protected NIOServerCnxn createConnection(SocketChannel sock, SelectionKey sk, SelectorThread selectorThread) throws IOException {
-        return new NIOServerCnxn(zkServer, sock, sk, this, selectorThread);
+        if (zkServer != null) {
+            return new NIOServerCnxn(zkServer, sock, sk, this, selectorThread);
+        }
+
+        LOG.warn("leader has not been voted yet, please wait a little time.");
+        throw new IOException("zkServer is null, chanel can't be used now.");
     }
 
```

Comment:
@yanllearnn 
- Yes, `zkServer` can be null(`Leader#shutdown ---> self.setZooKeeperServer(null) `), but I guess that we cannot get null pointer exception. 
- The workflow of code:
```
Leader#shutdown ---> self.setZooKeeperServer(null) ---> NIOServerCnxnFactory.SelectorThread#run ---> processAcceptedConnections ---> createConnection(zkServer=null) 

# when the client try to connect that uninitialized server
NIOServerCnxn#doIO---> NIOServerCnxn#readLength ---> if (!isZKServerRunning()) ---> throw new IOException("ZooKeeperServer not running")
```

Reply:
@maoling 
I think adding the check is good, despite what you have said.